lumicks/pylake/kymotracker/detail/msd_estimation.py

```diff
@@ -447,3 +469,274 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
+
+    References
+    ----------
+    .. [11] Vestergaard, C. L., Blainey, P. C., & Flyvbjerg, H. (2014). Optimal estimation of
+            diffusion coefficients from single-particle trajectories. Physical Review E, 89(2),
+            022726.
+    .. [12] Vestergaard, C. L. (2016). Optimizing experimental parameters for tracking of diffusing
+            particles. Physical Review E, 94(2), 022401.
+    """
+    if not 0 <= blur_constant <= 0.25:
+        raise ValueError("Motion blur constant should be between 0 and 1/4")
+
+    if len(x) < 3:
+        raise RuntimeError("Insufficient intervals for using the CVE")
+
+    # Determine the average step size
+    average_frame_step = np.mean(np.diff(frame_indices))
+    avg_dt = average_frame_step * dt
+
+    dx = np.diff(x)
+    mean_dx_squared = np.mean(dx**2)
+
+    if not variance_loc:
+        # Estimate the variance based on this frame
+        mean_dx_consecutive = np.mean(dx[1:] * dx[:-1])
+        # Equation 21 from [12] adapted for 1D.
+        diffusion_constant = mean_dx_squared / (2 * avg_dt) + mean_dx_consecutive / avg_dt
+        # Equation 15 from [11]. Note that static loc uncertainty is not affected by frame skipping.
+        variance_loc = (
+            blur_constant * mean_dx_squared + (2 * blur_constant - 1) * mean_dx_consecutive
+        )
+        var_diffusion = _var_cve_unknown_var(
+            diffusion_constant, variance_loc, dt, len(x), blur_constant, average_frame_step
+        )
+    else:
+        # We know the variance in advance. Equation 23 from [12] adapted for 1D.
+        diffusion_constant = (mean_dx_squared - 2.0 * variance_loc) / (
+            2.0 * (avg_dt - 2.0 * blur_constant * dt)
+        )
+        var_diffusion = _var_cve_known_var(
+            diffusion_constant,
+            variance_loc,
+            variance_variance_loc,
+            dt,
+            len(x),
+            blur_constant,
+            average_frame_step,
+        )
+
+    return diffusion_constant, var_diffusion, variance_loc
+
+
+def estimate_diffusion_cve(
+    frame_idx: npt.ArrayLike,
+    coordinate: npt.ArrayLike,
+    dt: float,
+    blur_constant: float,
+    unit: str,
+    unit_label: str,
+) -> DiffusionEstimate:
+    """Estimate diffusion constant based on covariance estimator
+
+    This function estimates the diffusion constant based on the covariance estimator and packs
+    the result in a `DiffusionEstimate`. See the docstring for `_cve` for more information.
+
+    Parameters
+    ----------
+    frame_idx : array_like
+        Frame indices
+    coordinate : array_like
+        Positions over time
+    dt : float
+        Time step
+    blur_constant : float
+        Motion blur coefficient. Should be between 0 and 1/4.
+    unit : str
+        Unit that the diffusion constant is specified in.
+    unit_label : str
+        Unit in TeX format used for plotting labels.
+    """
+
+    # We hardcode the blur constant for confocal for now (no motion blur)
```

Comment:
Hm, I would probably move this comment into the calling function of `kymotrack` (see comment below).

Reply:
Yeah... hmm... I wonder why I put it here. My guess is that I moved the =0 upstream at some point during writing this and forgot to move the comment with. Anyhow, good catch! 👍